### PR TITLE
[NOD-863] Write interface tests for the new database

### PR DESCRIPTION
**Pull request by stasatdaglabs**

### CODE REVIEW

File: database/cursor_test.go [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+		t.Fatalf("%s: Close "+
+			"unexpectedly failed: %s", testName, err)
+	}
+
+	// We expect First to return false
+	result := cursor.First()
+	if result {
+		t.Fatalf("%s: First "+
+			"unexpectedly returned true", testName)
+	}
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply database/cursor_test.go:276]: Created a ticket.

File: database/cursor_test.go [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+
+	// We expect Next to return false
+	result = cursor.Next()
+	if result {
+		t.Fatalf("%s: Next "+
+			"unexpectedly returned true", testName)
+	}
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply database/cursor_test.go:283]: Created a ticket.

File: database/cursor_test.go [hunk 3]
```diff
@@ -0,0 +1,284 @@
+// All tests within this file should call testForAllDatabaseTypes
+// over the actual test. This is to make sure that all supported
+// database types adhere to the interfaces defined in this package.
```
Comment (svarogg) [inline database/cursor_test.go:3]: To the interface assumptions

File: database/cursor_test.go [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+		t.Fatalf("%s: cursor unexpectedly "+
+			"returned true after being closed", testName)
+	}
+}
+
+func TestCursorFirst(t *testing.T) {
+	testForAllDatabaseTypes(t, "TestCursorFirst", testCursorFirst)
+}
+
+func testCursorFirst(t *testing.T, db database.Database, testName string) {
```
Comment (svarogg) [inline database/cursor_test.go:88]: Please add to this test/somewhere else/new test that exhausts a cursor, calls first, then makes sure next works again

File: database/cursor_test.go [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+		t.Fatalf("%s: Cursor unexpectedly "+
+			"returned true", testName)
+	}
+}
+
+func TestCursorSeek(t *testing.T) {
+	testForAllDatabaseTypes(t, "TestCursorSeek", testCursorSeek)
+}
+
+func testCursorSeek(t *testing.T, db database.Database, testName string) {
```
Comment (svarogg) [inline database/cursor_test.go:145]: Please add test that makes sure that after seek order is preseved (a.k.a. seek to 5th entry, call next, make sure that you are now on 6th entry)

File: database/database_test.go [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+func testDatabasePut(t *testing.T, db database.Database, testName string) {
+	// Put value1 into the database
+	key := database.MakeBucket().Key([]byte("key"))
+	value1 := []byte("value1")
+	err := db.Put(key, value1)
+	if err != nil {
+		t.Fatalf("%s: Put "+
+			"unexpectedly failed: %s", testName, err)
+	}
+
```
Comment (svarogg) [inline database/database_test.go:26]: Don't you want to check that get returns value1 now?

File: database/transaction_test.go [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+			defer teardownFunc()
+
+			testName := fmt.Sprintf("%s: TestTransactionGet", name)
+			testTransactionGet(t, db, testName)
+		}()
+	}
+}
+
+func testTransactionGet(t *testing.T, db database.Database, testName string) {
+	// Put a value into the database
```
[the remark this answers is not in the document]
Comment (svarogg) [reply database/transaction_test.go:74]: Also: modify something in tx (without commiting yet), verify it was not modified outside.
Then commit, and verify it was modified outside

**Comment by someone235** [comment]
Don't you wanna add TestTransactionRollbackUnlessClosed?

File: database/transaction_test.go [hunk 8]
```diff
@@ -337,3 +359,191 @@ func testTransactionAppendToStoreAndRetrieveFromStore(t *testing.T, db database.
...
+
+	// Put a value into the transaction
+	key := database.MakeBucket().Key([]byte("key"))
+	value := []byte("value")
+	err = dbTx.Put(key, value)
+	if err != nil {
+		t.Fatalf("%s: Put "+
+			"unexpectedly failed: %s", testName, err)
+	}
+
```
Comment (svarogg) [inline database/transaction_test.go:390]: Please also do db.Get("key") before commit, to make absolutely sure it wasn't applied before calling Commit

**Pull request merged**